File: src/test/java/com/integreety/yatspec/e2e/interceptor/LsdRestTemplateCustomizerShould.java
package com.integreety.yatspec.e2e.interceptor;

import com.integreety.yatspec.e2e.captor.http.RequestCaptor;
import com.integreety.yatspec.e2e.captor.http.ResponseCaptor;
import com.integreety.yatspec.e2e.captor.repository.model.InterceptedInteraction;
import org.junit.jupiter.api.Test;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.web.client.RestTemplate;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;

class LsdRestTemplateCustomizerShould {

    private final RequestCaptor requestCaptor = mock(RequestCaptor.class);
    private final ResponseCaptor responseCaptor = mock(ResponseCaptor.class);

    private final RestTemplate restTemplate = new RestTemplate();
    private final LsdRestTemplateInterceptor lsdRestTemplateInterceptor = new LsdRestTemplateInterceptor(requestCaptor, responseCaptor);

    private final LsdRestTemplateCustomizer underTest = new LsdRestTemplateCustomizer(lsdRestTemplateInterceptor);

    @Test
    void addsLsdInterceptor() {
        underTest.customize(restTemplate);

        assertThat(restTemplate.getInterceptors()).containsExactly(lsdRestTemplateInterceptor);
    }

    @Test
    void preservesExistingCustomizers() {
        restTemplate.setInterceptors(List.of(mock(ClientHttpRequestInterceptor.class)));

        underTest.customize(restTemplate);

        assertThat(restTemplate.getInterceptors()).hasSize(2);
    }

    @Test
    void doesntAddDuplicateInterceptor() {
        restTemplate.setInterceptors(List.of(lsdRestTemplateInterceptor));

        underTest.customize(restTemplate);

        assertThat(restTemplate.getInterceptors()).hasSize(1);
    }

    @Test
    void responseIsNotEmptyAfterInterception() {
        given(requestCaptor.captureRequestInteraction(any(), anyString())).willReturn(InterceptedInteraction.builder().build());

        underTest.customize(restTemplate);

        final String forObject = restTemplate.getForObject("https://httpbin.org/get", String.class);

        assertThat(forObject).isNotEmpty();
    }
}

File: spring-web/src/test/kotlin/io/lsdconsulting/lsd/distributed/interceptor/interceptor/LsdRestTemplateCustomizerShould.kt
package io.lsdconsulting.lsd.distributed.interceptor.interceptor

import io.github.krandom.KRandom
import io.github.krandom.KRandomParameters
import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.interceptor.captor.RequestCaptor
import io.lsdconsulting.lsd.distributed.interceptor.captor.ResponseCaptor
import io.mockk.every
import io.mockk.mockk
import org.hamcrest.CoreMatchers.nullValue
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.Test
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.web.client.RestTemplate

internal class LsdRestTemplateCustomizerShould {
    private val kRandom = KRandom(KRandomParameters().seed(System.currentTimeMillis()))
    private val requestCaptor = mockk<RequestCaptor>()
    private val responseCaptor = mockk<ResponseCaptor>()
    private val restTemplate = RestTemplate()
    private val lsdRestTemplateInterceptor = LsdRestTemplateInterceptor(requestCaptor, responseCaptor)

    private val underTest = LsdRestTemplateCustomizer(lsdRestTemplateInterceptor)

    @Test
    fun addsLsdInterceptor() {
        underTest.customize(restTemplate)

        assertThat(restTemplate.interceptors, contains(lsdRestTemplateInterceptor))
    }

    @Test
    fun `preserves existing customizers`() {
        restTemplate.interceptors = listOf(mockk())

        underTest.customize(restTemplate)

        assertThat(restTemplate.interceptors, hasSize(2))
    }

    @Test
    fun `doesnt add duplicate interceptor`() {
        restTemplate.interceptors = listOf<ClientHttpRequestInterceptor>(lsdRestTemplateInterceptor)

        underTest.customize(restTemplate)

        assertThat(restTemplate.interceptors, hasSize(1))
    }

    @Test
    fun `response is not empty after interception`() {
        every { requestCaptor.captureRequestInteraction(any(), any()) } returns kRandom.nextObject(InterceptedInteraction::class.java)

        underTest.customize(restTemplate)

        val forObject = restTemplate.getForObject("https://bbc.co.uk", String::class.java)
        assertThat(forObject, not(nullValue()))
    }
}
